### PR TITLE
Issue 12

### PR DESCRIPTION
**Pull request by phadej**
Resolves: #12 

ping  @fizruk, can you glance thru.

### CODE REVIEW

File: src/Data/HashMap/Strict/InsOrd.hs [hunk 1]
```diff
@@ -351,8 +360,11 @@ unionWith
     => (v -> v -> v)
     -> InsOrdHashMap k v -> InsOrdHashMap k v -> InsOrdHashMap k v
 unionWith f (InsOrdHashMap i a) (InsOrdHashMap j b) =
-    InsOrdHashMap (i + j) $ HashMap.unionWith f' a b'
+    mk $ HashMap.unionWith f' a b'
   where
+    -- the threshold is arbitrary, it meant to amortise need for packing of indices
+    mk | i + j > 0xffff = fromHashMapP
+       | otherwise      = InsOrdHashMap (i + j)
     b' = fmap (incPK i) b
```
Comment (fizruk) [inline src/Data/HashMap/Strict/InsOrd.hs]: Overflow can still happen here, right?

File: src/Data/HashMap/Strict/InsOrd.hs [hunk 2]
```diff
@@ -400,8 +412,65 @@ mapWithKey f (InsOrdHashMap i m) =
   where
     f' k (P j x) = P j (f k x)
 
+foldMapWithKey :: Monoid m => (k -> a -> m) -> InsOrdHashMap k a -> m
+foldMapWithKey f = foldMap (uncurry f) . toList
```
Comment (fizruk) [inline src/Data/HashMap/Strict/InsOrd.hs]: Why not just use `ifoldMap` here? If this one is better, why not define `ifoldMap = foldMapWithKey`?

File: src/Data/HashMap/Strict/InsOrd.hs [hunk 3]
```diff
@@ -400,8 +412,65 @@ mapWithKey f (InsOrdHashMap i m) =
...
+
 traverseWithKey :: Applicative f => (k -> a -> f b) -> InsOrdHashMap k a -> f (InsOrdHashMap k b)
-traverseWithKey f (InsOrdHashMap i m) =
+traverseWithKey f (InsOrdHashMap n m) = InsOrdHashMap n <$> retractSortedAp
+    (HashMap.traverseWithKey (\k (P i v) -> liftSortedAp i (P i <$> f k v)) m)
+
+-- Sort using insertion sort
```
Comment (fizruk) [inline src/Data/HashMap/Strict/InsOrd.hs]: Why not include it into Haddock documentation?

File: src/Data/HashMap/Strict/InsOrd.hs [hunk 4]
```diff
@@ -400,8 +412,65 @@ mapWithKey f (InsOrdHashMap i m) =
...
+-- Hopefully it's fast enough for where we need it
+-- otherwise: https://gist.github.com/treeowl/9621f58d55fe0c4f9162be0e074b1b29
+-- http://elvishjerricco.github.io/2017/03/23/applicative-sorting.html also related
+
+-- Free applicative which re-orders effects
+-- Mostly from Edward Kmett's `free` package.
```
Comment (fizruk) [inline src/Data/HashMap/Strict/InsOrd.hs]: `free` is a transitive dependency of `insert-ordered-containers`. We might use them directly! Except there's no `SortedAp` defined there at this time. Maybe we should add one there?

Also, `SortedAp f` seems to be like `Ap (Compose P f)` except with sorting built into `<*>`.
It looks like we can use just `Ap` together with something like:

```haskell
sortAp :: Ord1 f => Ap f a -> Ap f a
sortApWith :: (forall a. f a -> f a -> Ordering) -> Ap f b -> Ap f b

-- this instance is not ideal
instance Ord1 (Compose P f) where
  compare1 (Compose (P i _)) (Compose (P j _)) = compare1 i j
```

File: src/Data/HashMap/Strict/InsOrd.hs [hunk 5]
```diff
@@ -400,8 +412,65 @@ mapWithKey f (InsOrdHashMap i m) =
...
+    f@(SortedAp i x y) <*> z@(SortedAp j u v)
+        | i < j     = SortedAp i x (flip <$> y <*> z)
+        | otherwise = SortedAp j u ((.) <$> f <*> v)
+
+liftSortedAp :: Int -> f a -> SortedAp f a
+liftSortedAp i x = SortedAp i x (Pure id)
+
+retractSortedAp :: Applicative f => SortedAp f a -> f a
+retractSortedAp (Pure x) = pure x
+retractSortedAp (SortedAp _ f x) = flip id <$> f <*> retractSortedAp x
```
Comment (phadej) [inline src/Data/HashMap/Strict/InsOrd.hs]: I  just copy pasted this from `free` :)

File: src/Data/HashMap/Strict/InsOrd.hs (hunk 2 above)
Comment (phadej) [inline src/Data/HashMap/Strict/InsOrd.hs]: I was planning to use this to define `FoldableWithIndex`, good catch!

File: src/Data/HashMap/Strict/InsOrd.hs (hunk 3 above)
Comment (phadej) [inline src/Data/HashMap/Strict/InsOrd.hs]: SortedAp is not exported

**Comment by phadej** [comment]
Thanks for the review! I addressed the comments. Will release tomorrow if you or someone else doesn't find anything else wrong here.

**Pull request merged**